Sources/Experiment/ExperimentPlugin.swift

```diff
@@ -0,0 +1,230 @@
+            } else {
+                return nil
+            }
+        }
+    }
+
+    // Dedup instances by deployment key.
+    // A nil deployment key means we should use the API key of the analytics client we attach to.
+    // But this may not be available as the client may not be attached
+    // So we indicate this by deduping just on pluginName, as api key should be consistent
+    // per analytics client
+    fileprivate static func instanceName(deploymentKey: String?) -> String {
+        if let deploymentKey {
+            return pluginName + "_" + deploymentKey
+        } else {
+            return pluginName
+        }
+    }
+
+    @objc public init(config: Config = .init()) {
+        mode = .hosted(config: config)
+    }
+
+    @objc public init(experiment: ExperimentClient) {
+        mode = .external
+        self.experiment = experiment
+    }
+
+    public func setup(analyticsClient: any AmplitudeCore.AnalyticsClient,
+                      amplitudeContext: AmplitudeCore.AmplitudeContext) {
+        self.context = amplitudeContext
+        self.analytics = analyticsClient
+
+        switch mode {
+        case .hosted(let config):
+            let configBuilder = ExperimentConfigBuilder()
+            configBuilder.debug = config.debug
+            configBuilder.instanceName = amplitudeContext.instanceName
+            configBuilder.fallbackVariant = config.fallbackVariant
+            configBuilder.initialFlags = config.initialFlags
+            configBuilder.initialVariants = config.initialVariants
+            configBuilder.source = config.source
+            configBuilder.serverUrl = config.serverUrl
+            configBuilder.flagsServerUrl = config.flagsServerUrl
+            let serverZone: ServerZone
+            switch amplitudeContext.serverZone {
+            case .US:
+                serverZone = .US
+            case .EU:
+                serverZone = .EU
+            @unknown default:
+                serverZone = .US
+            }
+            configBuilder.serverZone = serverZone
+            configBuilder.fetchTimeoutMillis = config.fetchTimeoutMillis
+            configBuilder.retryFetchOnFailure = config.retryFetchOnFailure
+            configBuilder.automaticExposureTracking = config.automaticExposureTracking
+            configBuilder.fetchOnStart = config.fetchOnStart
+            configBuilder.pollOnStart = config.pollOnStart
+            configBuilder.flagConfigPollingIntervalMillis = config.flagConfigPollingIntervalMillis
+            configBuilder.automaticFetchOnAmplitudeIdentityChange = config.automaticFetchOnAmplitudeIdentityChange
+            configBuilder.userProvider = self
+            configBuilder.exposureTrackingProvider = self
+
+            experiment = Experiment.initialize(apiKey: config.deploymentKey ?? amplitudeContext.apiKey,
+                                               config: configBuilder.build())
+        case .external:
+            break
```

Comment:
Is there any action we'd want to take when attaching an instance of Amplitude to an instance of experiment?

Reply:
I'm not sure I fully understand the question. Could you clarify or provide an example?

Reply:
For an externally hosted experiment client, this is the first point in which we'd have both the analytics and experiment clients both set up and ready to go.

We could potentially:
- set `userProvider`/`exposureTrackingProvider` to the amplitude instance the plugin is attached to (if not already set?)
- call `setUser` to identify the user with Amplitude data.